File: 3647-zero-array-transformation-iii/zero-array-transformation-iii.java
class Solution {
    public int maxRemoval(int[] nums, int[][] queries) {
        int n = nums.length;
        int qz = queries.length;

        // Lưu danh sách các truy vấn kết thúc tại chỉ số nào
        List<Integer>[] qEnd = new ArrayList[n];
        for (int i = 0; i < n; i++) qEnd[i] = new ArrayList<>();

        for (int[] query : queries) {
            int l = query[0], r = query[1];
            qEnd[l].add(r);
        }

        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder()); // max heap
        int[] cntQ = new int[n + 1]; // số lần giảm đang hoạt động tại mỗi vị trí
        int dec = 0;

        for (int i = 0; i < n; i++) {
            dec += cntQ[i];

            // Thêm các truy vấn bắt đầu tại i vào heap
            for (int r : qEnd[i]) {
                pq.offer(r);
            }

            int x = nums[i];
            // Dùng các truy vấn trong heap để giảm nums[i] về 0
            while (x > dec && !pq.isEmpty() && pq.peek() >= i) {
                int k = pq.poll();
                cntQ[k + 1]--; // giảm lại tại điểm kết thúc + 1
                dec++;
            }

            if (x > dec) return -1; // Không thể giảm về 0
        }

        return pq.size(); // số truy vấn còn lại chưa dùng
    }
}

File: 3647-zero-array-transformation-iii/zero-array-transformation-iii.go
type MaxHeap []int

func (h MaxHeap) Len() int            { return len(h) }
func (h MaxHeap) Less(i, j int) bool  { return h[i] > h[j] } // max heap
func (h MaxHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func maxRemoval(nums []int, queries [][]int) int {
	n := len(nums)
	qEnd := make([][]int, n)
	for _, q := range queries {
		l, r := q[0], q[1]
		qEnd[l] = append(qEnd[l], r)
	}

	cntQ := make([]int, n+1)
	dec := 0
	pq := &MaxHeap{}
	heap.Init(pq)

	for i := 0; i < n; i++ {
		dec += cntQ[i]

		// Thêm các truy vấn bắt đầu tại i
		for _, r := range qEnd[i] {
			heap.Push(pq, r)
		}

		x := nums[i]
		for x > dec && pq.Len() > 0 && (*pq)[0] >= i {
			k := heap.Pop(pq).(int)
			cntQ[k+1]--
			dec++
		}

		if x > dec {
			return -1
		}
	}

	return pq.Len()
}